text/0000-Bundle-Load-API.md

```diff
@@ -1,63 +1,119 @@
+However, this not possible as DS runtime is designed such that it only loads the service class constructors/destructors by calling extern "C" functions instead of framework's Start function.
+As such, framework's Start API can not solve this specific use case for DS.
 
-seems like repetition and can be consolidated at one place, i.e in Bundle class.
-Given this it seems appropriate to have a Bundle::Load API for CppMicroServices library.
+Because of these reasons, DS uses OS specific dlopen/LoadLibraryW calls for loading bundles/shared objects.
+
+This loading of libraries/bundles increases the overhead in the declarative services while doing the symbol resoution and ref count increment done in the system.
+
+Adding to that, this functionality to load bundles/shared objects is currently implemented at two places
+
+1) Declarative services runtime
+2) framework
+
+During the application run, these different callsites may become out of sync.
+
+Given this it seems appropriate to have a Bundle::BundleLoad API for CppMicroServices library.
+
+## Requirements
+
+For the specific use case of DS (and similar extender clients) ,
+
+1) implement Bundle::LoadBundle such that it encapsulates shared library loading and symbol resolution, not exposing OS specific data types
+2) become a single function in the CppMicroServices eco system that loads the bundles/shared libraries
+3) reduce the amount of OS specific code a client needs to write
 
 ## Detailed design
 
-Bundle::Load() API will be declared in the file CppMicroServices/framework/include/cppmicroservices/Bundle.h as below 
+Bundle::LoadBundle() API will be declared in the file CppMicroServices/framework/include/cppmicroservices/Bundle.h as below
+The output of above API will be in the wrapper class which will be as belows.
 
 ```
 /**
-* Loads this bundle using platform APIs 
-* On POSIX   : dlopen
-* On Windows : LoadLibraryW
+* @brief  This class holds the output of BundleLoad API.
+*
+* It acts as a wrapper over the handles returned by the OS APIs that loads the library.
+* Clients have to supply the function type as template argument to correctly get the desired symbol resolved using
+* GetSymbol function. See the usage workflow for more details.
+*
+*/
+
+class BundleHandle
+{
+    public:
+        BundleHandle():handle(nullptr) {}
+        ~BundleHandle() { handle = nullptr; }
+        BundleHandle(void *handle) : handle(symHandle) {}
+
+        template<typename T>
+        T Getsymbol(const std::string& symname);
+
+    private:
+        void * symHandle;
+}
+
+/**
+* Loads this bundle using platform APIs and return a tuple of BundleSymbol class object
+* It uses following OS APIs to achieve the same
+*
+* On POSIX   : dlopen/dlsym
+* On Windows : LoadLibraryW/GetProcAddress
 * 
-* @return Handle to the bundle in the form of void *
+* A framework event of type FRAMEWORK_ERROR will fired with following information
```

Comment:
@jeffdiclemente Code inspection (and design RFC) did not give me any results if DS runtime handles the FramworkEvent at all. 
Given this, I am thinking of settling back to regular exception(s). Will that be ok ? Or should I read/dig more on this ? 

After your reply , I will update the scenarios in which exception/FrameworkEvent will fired.

Reply:
It will have to be decided what DS does with the exception and ```FrameworkEvent```, which I don't believe is in scope for this RFC.

Reply:
In that case, I think I will stick to exception for now and have a separate task for FrameworkEvent, will that be ok ?

Reply:
I think the decision on whether or not to send a FrameworkEvent is overshadowed by Sascha's questions:

- What use case does this proposal serve which Bundle::Start cannot today?
- Why is modifying Bundle::Start not an option?
- What makes DS use native "load library" calls instead of Bundle::Start?

once these questions have been answered, we can look into implementation details since the answers to these questions may change the implementation details.

Its time to take a step back and consider the problem we're trying to solve. I believe we are too focused on this ```LoadBundle``` solution.